packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -37,59 +39,131 @@ impl HookContextImpl {
 ///
-/// Snippets can be added to the current output via [`attach_snippet()`].
+/// Lines to the rendered tree of objects can be attached via [`HookContext::emit`], or
+/// [`HookContext::emit_deferred`]. Snippets can be added to the current output via
+/// [`HookContext::snippet`], or [`HookContext::snippet_deferred`].
 ///
-/// [`attach_snippet()`]: HookContext::attach_snippet
 /// [`Debug`]: core::fmt::Debug
 ///
+/// ### `emit`/`snippet` vs. `emit_deferred`/`snippet_deferred`
+///
+/// Lines and snippets can be emitted either immediately or deferred until the current stack, where
+/// a stack is a list of attachments until a frame which has more than a single source.
+///
+/// To emit a line immediately use [`HookContext::emit`] and [`HookContext::snippet`] for snippets,
+/// use [`HookContext::emit_deferred`] for deferred lines and [`HookContext::snippet_deferred`] for
+/// snippets.
+///
 /// ### Example
 ///
+/// TODO: add example of `snippets` deferred
 /// ```rust
 /// # // we only test with Rust 1.65, which means that `render()` is unused on earlier version
 /// # #![cfg_attr(not(rust_1_65), allow(dead_code, unused_variables, unused_imports))]
-/// use std::io::ErrorKind;
+/// use std::io::{Error, ErrorKind};
 ///
-/// use error_stack::{fmt::Emit, Report};
+/// use error_stack::Report;
 ///
-/// struct Error {
-///     code: usize,
-///     reason: &'static str,
-/// }
+/// struct Warning(&'static str);
+/// struct HttpResponseStatusCode(u64);
+/// struct Suggestion(&'static str);
+/// struct Secret(&'static str);
 ///
-/// Report::install_debug_hook::<Error>(|Error { code, reason }, ctx| {
+/// // You can emit a line, which is going to be emitted immediately.
+/// Report::install_debug_hook::<HttpResponseStatusCode>(|HttpResponseStatusCode(val), ctx| {
+///     // This is going to be emitted immediately in the list of snippets.
 ///     if ctx.alternate() {
-///         ctx.attach_snippet(format!("Error {code}:\n  {reason}"));
+///         ctx.snippet(format!("Error {val}: {} Error", if *val < 500 {"Client"} else {"Server"}))
 ///     }
 ///
-///     vec![Emit::next(format!("Error {code}"))]
+///     ctx.emit(format!("Error Code: {val}"));
 /// });
 ///
-/// let report = Report::new(std::io::Error::from(ErrorKind::InvalidInput))
-///     .attach(Error {
-///         code: 404,
-///         reason: "Not Found - Server cannot find requested resource",
-///     })
-///     .attach(Error {
-///         code: 405,
-///         reason: "Bad Request - Server cannot or will not process request",
-///     });
+/// // You can emit a line which is deferred to the end of the stack.
+/// Report::install_debug_hook::<Suggestion>(|Suggestion(val), ctx| {
+///     let idx = ctx.increment();
+///
+///     // deferred snippets are deferred until the end of snippets of the current stack
+///     if ctx.alternate() {
+///         ctx.snippet_deferred(format!("Suggestion {idx}:\n  {val}"));
+///     }
```

Comment:
This is not shown in the provided output. I'm fine with that, just want to flag in case you want to change this.

Reply:
fixed in [7640562](https://github.com/hashintel/hash/pull/1008/commits/76405629a6351a476bb7a4be3a1591fdc26ebca2)